x-pack/plugins/spaces/server/new_platform/spaces_service/spaces_service.ts

```diff
@@ -0,0 +1,110 @@
+import { first } from 'rxjs/operators';
+import { DEFAULT_SPACE_ID } from '../../../common/constants';
+import { SecurityPlugin } from '../../../../security';
+import { SpacesClient } from '../../lib/spaces_client';
+import { getSpaceIdFromPath } from '../../lib/spaces_url_parser';
+
+interface RequestFacade {
+  headers?: Headers;
+  getBasePath: () => string;
+}
+
+export interface SpacesServiceSetup {
+  scopedClient(request: RequestFacade): SpacesClient;
+
+  getSpaceId(request: RequestFacade): string;
+
+  isInDefaultSpace(request: RequestFacade): boolean;
+}
+
+interface CacheEntry {
+  spaceId: string;
+  isInDefaultSpace: boolean;
+}
+
+interface SpacesServiceDeps {
+  elasticsearch: ElasticsearchServiceSetup;
+  savedObjects: SavedObjectsService;
+  getSecurity: () => SecurityPlugin;
+  spacesAuditLogger: any;
+}
+
+export class SpacesService {
+  private readonly serverBasePath: string;
+
+  private readonly contextCache: WeakMap<any, CacheEntry>;
+
+  constructor(private readonly log: Logger, private readonly config: KibanaConfig) {
+    this.serverBasePath = config.get('server.basePath');
+
+    this.contextCache = new WeakMap();
+  }
+
+  public async setup({
+    elasticsearch,
+    savedObjects,
+    getSecurity,
+    spacesAuditLogger,
+  }: SpacesServiceDeps): Promise<SpacesServiceSetup> {
+    const adminClient = await elasticsearch.adminClient$.pipe(first()).toPromise();
+    return {
+      getSpaceId: (request: RequestFacade) => {
+        if (!this.contextCache.has(request)) {
+          this.populateCache(request);
+        }
+
+        const { spaceId } = this.contextCache.get(request) as CacheEntry;
+        return spaceId;
+      },
+      isInDefaultSpace: (request: RequestFacade) => {
+        if (!this.contextCache.has(request)) {
+          this.populateCache(request);
+        }
+
+        return this.contextCache.get(request)!.isInDefaultSpace;
+      },
+      scopedClient: (request: RequestFacade) => {
+        const internalRepository = savedObjects.getSavedObjectsRepository(
+          adminClient.callAsInternalUser
+        );
+
+        const callCluster = (endpoint: string, ...args: any[]) =>
+          adminClient.asScoped(request).callAsCurrentUser(endpoint, ...args);
+
+        const callWithRequestRepository = savedObjects.getSavedObjectsRepository(callCluster);
+
+        const security = getSecurity();
+        const authorization = security ? security.authorization : null;
+
+        return new SpacesClient(
+          spacesAuditLogger,
+          (message: string) => {
+            this.log.debug(message);
+          },
+          authorization,
+          callWithRequestRepository,
+          this.config,
+          internalRepository,
+          request
+        );
+      },
+    };
+  }
+
+  private populateCache(request: RequestFacade) {
+    const spaceId = getSpaceIdFromPath(request.getBasePath(), this.serverBasePath);
```

Comment:
I suspect that
- `getBasePath` can be a basic functionality provided by http or basePath service (need to discuss https://github.com/elastic/kibana/pull/35429/files#r279371736). base path value is based on incoming request url after parsing.
- `setBasePath` is implemented in Security plugin via registerOnRequest hook and redirects to a route with defined space
What do you think @legrego ?